sql/analyzer/indexed_joins.go

```diff
@@ -634,287 +677,214 @@ func dfsExprGroup(grp *exprGroup, m *Memo, seen map[GroupId]struct{}, cb func(re
+	// |grp| provides all tables referenced in |e|
+	return e.Group().ScalarProps().Tables.Difference(grp.RelProps.OutputTables()).Len() == 0
 }
 
 // addMergeJoins will add merge join operators to join relations
 // with native indexes providing sort enforcement on an equality
 // filter.
 // TODO: sort-merge joins
-func addMergeJoins(m *Memo) error {
+func addMergeJoins(m *memo.Memo) error {
 	var aliases = make(TableAliases)
-	seen := make(map[GroupId]struct{})
-	return dfsExprGroup(m.root, m, seen, func(e relExpr) error {
-		var join *joinBase
+	return dfsRel(m.Root(), func(e memo.RelExpr) error {
+		var join *memo.JoinBase
 		switch e := e.(type) {
-		case *innerJoin:
-			join = e.joinBase
-		case *leftJoin:
-			join = e.joinBase
+		case *memo.InnerJoin:
+			join = e.JoinBase
+		case *memo.LeftJoin:
+			join = e.JoinBase
 			//TODO semijoin, antijoin, fullouterjoin
 		default:
 			return nil
 		}
 
-		if len(join.filter) == 0 {
+		if len(join.Filter) == 0 {
 			return nil
 		}
 
-		lAttrSource, lIndexes, err := lookupCandidates(m.ctx, join.left.first, aliases)
+		lAttrSource, leftGrp, lIndexes, err := lookupCandidates(m.Ctx, join.Left.First, aliases)
 		if err != nil {
 			return err
 		} else if lAttrSource == "" {
 			return nil
 		}
-		rAttrSource, rIndexes, err := lookupCandidates(m.ctx, join.right.first, aliases)
+		rAttrSource, rightGrp, rIndexes, err := lookupCandidates(m.Ctx, join.Right.First, aliases)
 		if err != nil {
 			return err
-		} else if rAttrSource == "" {
-			return nil
 		}
 
-		for i, f := range join.filter {
-			var l, r sql.Expression
+		if tab, ok := aliases[lAttrSource]; ok {
+			lAttrSource = strings.ToLower(tab.Name())
+		}
+		if tab, ok := aliases[rAttrSource]; ok {
+			rAttrSource = strings.ToLower(tab.Name())
+		}
+
+		for i, f := range join.Filter {
+			var l, r *memo.ExprGroup
 			switch f := f.(type) {
-			case *expression.Equals:
-				l = f.Left()
-				r = f.Right()
+			case *memo.Equal:
+				l = f.Left
+				r = f.Right
 			default:
 				continue
 			}
 
-			ltc, ok := attrsRefSingleTableCol(l)
-			if !ok {
-				continue
-			}
-			rtc, ok := attrsRefSingleTableCol(r)
-			if !ok {
+			if l.ScalarProps().Cols.Len() != 1 ||
+				r.ScalarProps().Cols.Len() != 1 {
 				continue
 			}
 
-			switch {
-			case ltc.table == lAttrSource && rtc.table == rAttrSource:
-			case rtc.table == lAttrSource && ltc.table == rAttrSource:
+			var swap bool
+			if l.ScalarProps().Tables.Contains(int(memo.TableIdForSource(leftGrp))) &&
+				r.ScalarProps().Tables.Contains(int(memo.TableIdForSource(rightGrp))) {
+			} else if r.ScalarProps().Tables.Contains(int(memo.TableIdForSource(leftGrp))) &&
+				l.ScalarProps().Tables.Contains(int(memo.TableIdForSource(rightGrp))) {
+				swap = true
 				l, r = r, l
-				ltc, rtc = rtc, ltc
-			default:
+			} else {
 				continue
 			}
 
+			var lRef *memo.ColRef
+			dfsScalar(l.Scalar, func(e memo.ScalarExpr) error {
+				var err error
```

Comment:
Seems odd to define an error var outside the scope it's returned, rather than returning directly and returning nil for the final expression

In general I think this is kind of an anti-pattern, don't try to gather up error vars to return in exactly one place, return errors where they are generated unless you have a really good reason not to